prow/git/v2/interactor.go

```diff
@@ -52,3 +62,262 @@ type Interactor interface {
 	// MergeCommitsExistBetween determines if merge commits exist between target and HEAD
 	MergeCommitsExistBetween(target, head string) (bool, error)
 }
+
+// CacheCloner knows how to clone repositories and update them for the central cache
+type CacheCloner interface {
+	// Clone clones the repository from a local path.
+	Clone(from string) error
+	// MirrorClone sets up a mirror of the source repository.
+	MirrorClone() error
+	// RemoteUpdate fetches all updates from the remote.
+	RemoteUpdate() error
+}
+
+type interactor struct {
+	executor Executor
+	remote   RemoteResolver
+	dir      string
+	logger   *logrus.Entry
+}
+
+// Directory exposes the directory in which this repository has been cloned
+func (i *interactor) Directory() string {
+	return i.dir
+}
+
+// Clean cleans up the repository from the on-disk cache
+func (i *interactor) Clean() error {
+	return os.RemoveAll(i.dir)
+}
+
+// Clone clones the repository from a local path.
+func (i *interactor) Clone(from string) error {
+	i.logger.Infof("Creating a clone of the repo at %s from %s", i.dir, from)
+	if out, err := i.executor.Run("clone", from, i.dir); err != nil {
+		return fmt.Errorf("error creating a clone: %v %v", err, string(out))
+	}
+	return nil
+}
+
+// MirrorClone sets up a mirror of the source repository.
+func (i *interactor) MirrorClone() error {
+	i.logger.Infof("Creating a mirror of the repo at %s", i.dir)
+	remote, err := i.remote()
+	if err != nil {
+		return fmt.Errorf("could not resolve remote for cloning: %v", err)
+	}
+	if out, err := i.executor.Run("clone", "--mirror", remote, i.dir); err != nil {
+		return fmt.Errorf("error creating a mirror clone: %v %v", err, string(out))
+	}
+	return nil
+}
+
+// Checkout runs git checkout.
+func (i *interactor) Checkout(commitlike string) error {
+	i.logger.Infof("Checking out %q", commitlike)
+	if out, err := i.executor.Run("checkout", commitlike); err != nil {
+		return fmt.Errorf("error checking out %q: %v %v", commitlike, err, string(out))
+	}
+	return nil
+}
+
+// RevParse runs git rev-parse.
+func (i *interactor) RevParse(commitlike string) (string, error) {
+	i.logger.Infof("Parsing revision %q", commitlike)
+	out, err := i.executor.Run("rev-parse", commitlike)
+	if err != nil {
+		return "", fmt.Errorf("error parsing %q: %v %v", commitlike, err, string(out))
+	}
+	return string(out), nil
+}
+
+// BranchExists returns true if branch exists in heads.
+func (i *interactor) BranchExists(branch string) bool {
+	i.logger.Infof("Checking if branch %q exists", branch)
+	_, err := i.executor.Run("ls-remote", "--exit-code", "--heads", "origin", branch)
+	return err == nil
+}
+
+// CheckoutNewBranch creates a new branch and checks it out.
+func (i *interactor) CheckoutNewBranch(branch string) error {
+	i.logger.Infof("Checking out new branch %q", branch)
+	if out, err := i.executor.Run("checkout", "-b", branch); err != nil {
```

Comment:
This assumes the branch does not exist. We should either document it, or do `git branch --force; git checkout`.

Reply:
This simply achieves parity with the old client. I don't think we should change the way this works in this PR. Follow-ups after the `v1` client is removed could change this.

Reply:
Do we need to achieve bug parity, too? Here the fix is in implementation and not even that invasive (`git branch --force; git checkout`), or documentation.